File: Bookcase core/src/main/java/cz/vhromada/book/facade/BookFacade.java
package cz.vhromada.book.facade;

import cz.vhromada.book.entity.Book;
import cz.vhromada.common.facade.MovableParentFacade;
import cz.vhromada.result.Result;

/**
 * An interface represents facade for books.
 *
 * @author Vladimir Hromada
 */
public interface BookFacade extends MovableParentFacade<Book> {

    /**
     * Adds book. Sets new ID and position.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Book is null</li>
     * <li>ID isn't null</li>
     * <li>Position isn't null</li>
     * <li>Czech name is null</li>
     * <li>Czech name is empty string</li>
     * <li>Original name is null</li>
     * <li>Original name is empty string</li>
     * <li>Languages are null</li>
     * <li>Languages are empty</li>
     * <li>Languages contain null value</li>
     * <li>ISBN is empty string</li>
     * <li>Issue year isn't between 1940 and current year/</li>
     * <li>Description is null</li>
     * <li>Description is empty string</li>
     * <li>Note is empty string</li>
     * <li>Author ID is null</li>
     * <li>Author first name is empty string</li>
     * <li>Author middle name is empty string</li>
     * <li>Author last name is empty string</li>
     * <li>Author doesn't exist</li>
     * <li>Category ID is null</li>
     * <li>Category name is empty string</li>
     * <li>Category doesn't exist</li>
     * </ul>
     *
     * @param data book
     * @return result with validation errors
     */
    @Override
    Result<Void> add(Book data);

    /**
     * Updates book.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Book is null</li>
     * <li>ID is null</li>
     * <li>Position is null</li>
     * <li>Czech name is null</li>
     * <li>Czech name is empty string</li>
     * <li>Original name is null</li>
     * <li>Original name is empty string</li>
     * <li>Languages are null</li>
     * <li>Languages are empty</li>
     * <li>Languages contain null value</li>
     * <li>ISBN is empty string</li>
     * <li>Issue year isn't between 1940 and current year/</li>
     * <li>Description is null</li>
     * <li>Description is empty string</li>
     * <li>Note is empty string</li>
     * <li>Author ID is null</li>
     * <li>Author first name is empty string</li>
     * <li>Author middle name is empty string</li>
     * <li>Author last name is empty string</li>
     * <li>Author doesn't exist in data storage</li>
     * <li>Category ID is null</li>
     * <li>Category name is empty string</li>
     * <li>Category doesn't exist in data storage</li>
     * <li>Book doesn't exist in data storage</li>
     * </ul>
     *
     * @param data new value of book
     * @return result with validation errors
     */
    @Override
    Result<Void> update(Book data);

}

File: Bookcase core/src/main/kotlin/cz/vhromada/book/facade/BookFacade.kt
package cz.vhromada.book.facade

import cz.vhromada.book.entity.Book
import cz.vhromada.result.Result

/**
 * An interface represents facade for books.
 *
 * @author Vladimir Hromada
 */
interface BookFacade : BookcaseFacade<Book> {

    /**
     * Adds data. Sets new ID and position.
     *
     * Validation errors:
     * * ID isn't null
     * * Czech name is empty string
     * * Original name is empty string
     * * ISBN is empty string
     * * Issue year isn't between 1940 and current year
     * * Description is empty string
     * * Note is empty string
     * * Author ID is null
     * * Author first name is empty string
     * * Author middle name is empty string
     * * Author last name is empty string
     * * Author doesn't exist
     * * Category ID is null
     * * Category name is empty string
     * * Category doesn't exist
     *
     * @param data data
     * @return result with validation errors
     */
    override fun add(data: Book): Result<Void>

    /**
     * Updates data.
     *
     * Validation errors:
     * * ID is null
     * * Czech name is empty string
     * * Original name is empty string
     * * ISBN is empty string
     * * Issue year isn't between 1940 and current year
     * * Description is empty string
     * * Note is empty string
     * * Author ID is null
     * * Author first name is empty string
     * * Author middle name is empty string
     * * Author last name is empty string
     * * Author doesn't exist
     * * Category ID is null
     * * Category name is empty string
     * * Category doesn't exist
     * * Book doesn't exist in data storage
     *
     * @param data new value of data
     * @return result with validation errors
     */
    override fun update(data: Book): Result<Void>

}
